capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+                } else {
+                    // check for next data block to be ready
+                    self.state.set(State::WaitReadBlocks { count: count - 1 });
+                    self.read_bytes(write_buffer, read_buffer, 1);
+                }
+            }
+
+            State::ReadBlocksComplete => {
+                // check response
+                let (r1, _, _) = self.get_response(SDResponse::R1, read_buffer);
+
+                if r1 == 0x00 {
+                    // replace buffers
+                    self.txbuffer.replace(write_buffer);
+                    self.rxbuffer.replace(read_buffer);
+                    self.state.set(State::Idle);
+
+                    // read finished, perform callback
+                    self.client_buffer.take().map(move |buffer| {
+                        self.client.get().map(move |client| {
+                            client.read_done(buffer, self.client_offset.get());
+                        });
+                    });
+                } else {
+                    // error, send callback and quit
+                    self.txbuffer.replace(write_buffer);
+                    self.rxbuffer.replace(read_buffer);
+                    self.state.set(State::Idle);
+                    self.alarm_state.set(AlarmState::Idle);
+                    self.alarm_count.set(0);
+                    self.client
+                        .get()
+                        .map(move |client| { client.error(ErrorCode::ReadFailure as u32); });
+                }
+            }
+
+            State::StartWriteBlocks { count } => {
+                // check response
+                let (r1, _, _) = self.get_response(SDResponse::R1, read_buffer);
+
+                if r1 == 0x00 {
+                    if count <= 1 {
+                        // copy over data from client buffer
+                        let remaining_bytes = self.client_buffer.map_or(512, |buffer| {
+                            let write_len = cmp::min(buffer.len(), 512);
+
+                            for i in 0..write_len {
+                                write_buffer[i + 1] = buffer[i];
+                            }
+
+                            512 - write_len
+                        });
+
+                        // set a known value for remaining bytes
+                        for i in 0..remaining_bytes {
+                            write_buffer[i + 1] = 0xFF;
+                        }
+
+                        // set up data packet
+                        write_buffer[0] = 0xFE; // Data token
+                        write_buffer[513] = 0xFF; // dummy CRC
+                        write_buffer[514] = 0xFF; // dummy CRC
+
+                        // write data packet
+                        self.state.set(State::WriteBlockResponse);
+                        self.write_bytes(write_buffer, read_buffer, 515);
+                    } else {
+                        panic!("Multi-block SD card writes are unimplemented");
```

Comment:
I actually already do bubble up an ENOSUPPORT. In practice there is no way for SDCardDriver to reach this line of code